openwisp_firmware_upgrader/base/models.py

```diff
@@ -518,12 +519,29 @@ def log_line(self, line, save=True):
         if save:
             self.save()
 
+    def _recoverable_failure_handler(self, recoverable, error):
+        cause = str(error)
+        if recoverable:
+            self.log_line(f'Detected a recoverable failure: {cause}.\n', save=False)
+            self.log_line('The upgrade operation will be retried soon.')
+            raise error
+        self.status = 'failed'
+        self.log_line(f'Max retries exceeded. Upgrade failed: {cause}.', save=False)
+
     def upgrade(self, recoverable=True):
-        conn = self.device.deviceconnection_set.first()
-        installed = False
-        if not conn:
+        DeviceConnection = swapper.load_model('connection', 'DeviceConnection')
+        if self.device.deviceconnection_set.count() == 0:
             self.log_line('No device connection available')
             return
+        try:
+            conn = DeviceConnection.get_working_connection(self.device)
+        except NoWorkingDeviceConnectionError:
+            self._recoverable_failure_handler(
+                recoverable, RecoverableFailure('Connection failed')
```

Comment:
I think we need to explicitly state the credentials which were tried and the failure reason for each of them (one per line).
It's a bit redundant with the info available in the credentials tab but it helps user find this information faster when they're working on upgrades which is better.

Please add a test for this and in the test assert that `DeviceConnection.get_working_connection` is getting called.

Reply:
The `NoWorkingDeviceConnectionError` only contains information about the last tried credential. 


https://github.com/openwisp/openwisp-controller/blob/8481351812f06f9afee8e9936f57a2ce84398a8a/openwisp_controller/connection/base/models.py#L265